src/approximations/cvi.jl

```diff
@@ -25,26 +25,37 @@ Arguments
  - `n_samples`: number of samples to use for statistics approximation
  - `num_iterations`: number of iteration for the natural parameters gradient optimization
  - `opt`: optimizer, which will be used to perform the natural parameters gradient optimization step
+ - `grad`: optional, structure to select how the gradient and the hessian will be computed
  - `warn`: optional, defaults to false, enables or disables warnings related to the optimization steps
+ - `proper_message`: optional, defaults to true, enables or disables inforce for the approximation be a proper distribution
 
 !!! note 
     Run `using Flux` in your Julia session to enable the `Flux` optimizers support for the CVI approximation method.
 
+!!! note 
+    Run `using Zygote` in your Julia session to enable the `ZygoteGrad` option support for the CVI grad.
+
 """
-struct CVIApproximation{R, O} <: AbstractApproximationMethod
+struct CVIApproximation{R, O, G} <: AbstractApproximationMethod
     rng::R
     n_samples::Int
     num_iterations::Int
     opt::O
+    grad::G
     warn::Bool
+    proper_message::Bool
```

Comment:
I would make this field a `Val` based field, such that all checks could be compiled out:
https://docs.julialang.org/en/v1/manual/types/#%22Value-types%22

Reply:
@bvdmitri Could you elaborate on this a bit? Do you mean we will create a structure:
```julia
structure ProperMessage{x} end

ProperMessage(x::bool) = ProperMessage{x}()

enforce_proper_messages(::ProperMessage{false}, ...) = false
enforce_proper_messages(::ProperMessage{true}, ...) = isproper(message)
```?

Reply:
There is a built-in `Val{true}/Val{false}` for that purpose.

Reply:
[Done](https://github.com/biaslab/ReactiveMP.jl/pull/239/commits/7dfd256500c1c53446fb9d1c8e1722f43d2225e0)